src/lib/integration_test_lib/error_monad.ml

```diff
@@ -0,0 +1,167 @@
+open Core_kernel
+open Async_kernel
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Error.t
+          [@equal
+            fun a b ->
+              String.equal (Error.to_string_hum a) (Error.to_string_hum b)]
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    { computation_result: 'a
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module ErrorReporting = Monad.Make (struct
+  type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
+
+  let bind res ~f =
+    match%bind res with
+    | Ok {Accumulator.computation_result= prev_result; soft_errors= err_list}
+      -> (
+        let execProg = f prev_result in
+        match%bind execProg with
+        | Ok {Accumulator.computation_result= comp; soft_errors= next_list} ->
+            Deferred.return
+              (Ok
+                 { Accumulator.computation_result= comp
+                 ; Accumulator.soft_errors= List.append err_list next_list })
+        | Error
+            {Hard_fail.hard_error= hard_err; Hard_fail.soft_errors= next_list}
+          ->
+            Deferred.return
+              (Error
+                 { Hard_fail.hard_error= hard_err
+                 ; soft_errors= List.append err_list next_list }) )
+    | Error {Hard_fail.hard_error= hard_err; Hard_fail.soft_errors= err_list}
```

Comment:
You can clean this case up by changing it to `| Error _ as error -> Deferred.return error`.